### PR TITLE
[NavigationBar] Fix navbar placement in catalog for iPhoneX

### PR DESCRIPTION
**Pull request by yarneo**
Fixes issue #2580 

In the Catalog examples, I have taken into regards the safeAreaLayoutGuide introduced in iOS 11 so the iPhoneX will have the navigation bar in the right location.

### CODE REVIEW

File: components/NavigationBar/examples/NavigationBarLayoutExample.m [hunk 1]
```diff
@@ -106,15 +106,43 @@ - (void)viewDidLoad {
   [self.view addSubview:self.trailingItemField];
   [self.view addSubview:self.titleField];
 
+#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
+  if (@available(iOS 11.0, *)) {
+    [self.view.safeAreaLayoutGuide.topAnchor constraintEqualToAnchor:self.navigationBar.topAnchor].active = YES;
+  }
```
Comment (jverkoey) [inline components/NavigationBar/examples/NavigationBarLayoutExample.m]: What happens if we're building for iOS 11 but running on iOS 10? It looks like we're not adding any constraints in that case.

**Review by jverkoey — CHANGES_REQUESTED** [review]

**Comment by jverkoey** [comment]
Before/after screenshots please!

File: components/NavigationBar/examples/NavigationBarLayoutExample.m (hunk 1 above)
Comment (yarneo) [reply components/NavigationBar/examples/NavigationBarLayoutExample.m]: thanks! good call.

**Comment by yarneo** [comment]
iPhone 6 iOS 10.0
iPhone 6 iOS 11.0
iPhone X

**Review by jverkoey — APPROVED** [review]

**Comment by romoore** [comment]
Travis failed:

```
❌  /Users/travis/build/material-components/material-components-ios/components/NavigationBar/examples/NavigationBarTypicalUseExample.swift:41:14: value of type 'UIView' has no member 'safeAreaLayoutGuide'
        self.view.safeAreaLayoutGuide.topAnchor.constraint(equalTo: self.navBar!.topAnchor).isActive = true
```

File: components/NavigationBar/examples/NavigationBarIconsExample.m [hunk 2]
```diff
@@ -78,19 +78,21 @@ - (void)viewDidLoad {
...
-    [NSLayoutConstraint constraintWithItem:self.topLayoutGuide
-                                 attribute:NSLayoutAttributeBottom
-                                 relatedBy:NSLayoutRelationEqual
-                                    toItem:self.navigationBar
-                                 attribute:NSLayoutAttributeTop
-                                multiplier:1.0
-                                  constant:0]
-        .active = YES;
-  }
+  if ([[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){
```
Comment (romoore) [inline components/NavigationBar/examples/NavigationBarIconsExample.m]: This should be protected with a compile-time SDK check: 

```objectivec


#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
  if (@available(iOS 11.0, *)) {
    // iOS 11 code goes here
  } else {
#endif

  // iOS 8-10 code goes here

#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
  } // iOS 11 SDK
#endif
```

Comment (yarneo) [reply components/NavigationBar/examples/NavigationBarIconsExample.m]: Thanks @romoore will update accordingly.

File: components/NavigationBar/examples/NavigationBarTypicalUseExample.swift [hunk 3]
```diff
@@ -37,14 +37,34 @@ open class NavigationBarTypicalUseSwiftExample: NavigationBarTypicalUseExample {
 
     navBar!.translatesAutoresizingMaskIntoConstraints = false
 
-    let viewBindings = ["navBar": navBar!]
+    #if swift(>=3.2)
+      if #available(iOS 11.0, *) {
+        self.view.safeAreaLayoutGuide.topAnchor.constraint(equalTo: self.navBar!.topAnchor).isActive = true
+      } else {
+        NSLayoutConstraint(item: self.topLayoutGuide,
```
Comment (romoore) [inline components/NavigationBar/examples/NavigationBarTypicalUseExample.swift:44]: Can you use the same #if #endif as in Objc examples? It should avoid duplicating the non-iOS 11 constraints, yes?

**Review by romoore — APPROVED** [review]

File: components/NavigationBar/examples/NavigationBarTypicalUseExample.swift (hunk 3 above)
Comment (yarneo) [reply components/NavigationBar/examples/NavigationBarTypicalUseExample.swift:44]: I wish; Swift macros don't play as nice:
screen shot 2017-12-08 at 12 08 39 pm

**Comment by ianegordon** [comment]
Unless @romoore has any recommendations around the #if statements (agreed, not pretty), please resolve and merge.

File: components/NavigationBar/examples/NavigationBarTypicalUseExample.swift (hunk 3 above)
Comment (romoore) [reply components/NavigationBar/examples/NavigationBarTypicalUseExample.swift:44]: Must some sort of fake macros. #PreprocessorForever

**Pull request merged**